crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+
+                                // We're overwriting the first value in the block.
+                                self.inner[i] =
+                                    SliceHash::Single { id: slice_id, hash };
+
+                                if count > 1 {
+                                    self.inner.insert(
+                                        i + 1,
+                                        SliceHash::Block {
+                                            id: slice_id + 1,
+                                            count: count - 1,
+                                        },
+                                    );
+                                }
+                            } else if slice_id == id + count {
+                                // Special case for inserting at the end of the block.
+
+                                // We're overwriting the last value in the block.
+                                self.inner[i] =
+                                    SliceHash::Single { id: slice_id, hash };
+
+                                if count > 1 {
+                                    self.inner.insert(
+                                        i,
+                                        SliceHash::Block {
+                                            id,
+                                            count: count - 1,
+                                        },
+                                    );
+                                }
+                            } else {
+                                // Otherwise we're inserting in the middle of a block.
+
+                                // Contained within a compressed block but not empty, so we need
+                                // to split the block into two.
+                                let lower_block = SliceHash::Block {
+                                    id,
+                                    count: slice_id - id - 1,
+                                };
+                                let upper_block = SliceHash::Block {
+                                    id: slice_id + 1,
+                                    count: count - (slice_id - id) - 1,
+                                };
+
+                                // We're going to do one overwrite and two inserts.
+                                self.inner.reserve(2);
+                                self.inner[i] = lower_block;
+                                self.inner.insert(
+                                    i + 1,
+                                    SliceHash::Single { id: slice_id, hash },
+                                );
+                                self.inner.insert(i + 2, upper_block);
+                            }
+                        }
+
+                        handled = true;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if !handled {
```

Comment:
Thinking out loud: If our new hash isn't replacing an existing hash, and isn't being inserting upon an existing block, then shouldn't it get inserted to the end of the list, with a block inserted/extended before it? But I guess that is never the case since we're starting with a single block covering all time. So it seems like this error is unreachable. Is that right?

Reply:
Assuming the algorithm is correct, this is unreachable yes. However, because the loop isn't guaranteed to do work if there is an integrity problem with the data structure, I like having this here to catch the problem.

That does mean it isn't tested because I can't find a way to hit the condition. I'm happy to remove this is that would be preferred

Reply:
Now that I've used `rfind` below, it's much clearer what this `handled` was for and it's just part of a match condition after the find

Reply:
Also gone